### ISSUE TITLE
Secured endpoints do not work with OIDC

### ISSUE THREAD

**Issue by Tcharl** [jhipster/generator-jhipster#9735]
##### **Overview of the issue**

We systematically get a 401 when using `@Secured(AuthoritiesConstants.ADMIN)` with OIDC configuration.
The authorityMapper is not called, and also it's not accurate when compared to the JWT token.
What's the best way for fixing the issue?

##### **Motivation for or Use Case**

Securing endpoint

##### **Reproduce the error**

add a @Secured annotation on a resource endpoint (on a microservice on my side).
Try to call it, and dejoy... (tested on a microservice).
Also, I noticed that there were some agressive cast of Authentication to OAuth2AuthenticationToken, which is not always the case (for exemple service to service communication with Feign). this is not particularly related but to take into account while fixing. 

##### **Suggest a Fix**

Don't know, an AuthenticationManager implementation + a good implementation authorityMapper. Or maybe a token enhancer?

##### **JHipster Version(s)**

6.0.0

##### **JHipster configuration**

```
application {
  config {
    baseName pmi
    packageName com.corp.ecwp.pmi
    applicationType gateway
    authenticationType oauth2
    cacheProvider hazelcast
    clientFramework angularX
    clientPackageManager yarn
    enableHibernateCache true
    languages [en, fr]
    searchEngine elasticsearch
    serverPort 8080
    serviceDiscoveryType eureka
    testFrameworks [protractor]
    useSass true
  }
  entities X
}

application {
  config {
    baseName marketplace
    applicationType microservice
    authenticationType oauth2
    cacheProvider hazelcast
    clientPackageManager yarn
    databaseType sql
    enableHibernateCache true
    languages [en, fr]
    packageName com.corp.service.marketplace
    searchEngine elasticsearch
    serverPort 8103
    serviceDiscoveryType eureka
    useSass true
  }
  entities Y
}
```

-   [x] Checking this box is mandatory (this is just to show you read everything)

**Comment by antarus** [comment]
same problem here without use @Secured(AuthoritiesConstants.ADMIN)
If i comment the line
`.antMatchers("/api/**").authenticated()`
in `SecurityConfiguration.java` of my microservice, the service is accessible via the gateway
when is not commented, its try to redirect to`myentity/oauth2/authorization/oidc` and next 404 error page

**Comment by mraible** [comment]
Are you able to debug the JWT and look at its claims to ensure it has a `groups` claim? Either that or insect the current Spring Security user to see what their authorities are.

**Comment by Tcharl** [comment]
Yes, I debugged it and there's no `groups` claim at all.

See the new implemenation of `grantedauthoritymapper` and the `extractAuthorityFromClaims`  method I provided with https://github.com/jhipster/generator-jhipster/issues/9707

**Comment by mraible** [comment]
Can you please submit a PR to fix it?

**Comment by raulrotundo** [comment]
I added the "groups" claim to the token using keycloak and still having the AUTHORIZATION_FAILURE when a resource has been protected by @Secured or @PreAuthorize.

You can see some logs and comments by me in this article by @mraible https://developer.okta.com/blog/2019/05/23/java-microservices-spring-cloud-config#comment-4475357481

**Comment by vishal423** [comment]
I spent some time today on this and @Tcharl is correct. We need to provide a custom implementation of `JwtAuthenticationConverter` to be able to extract roles from access token (both on gateway and micro-services). It's required on gateway for resource server authentication to work correctly. I think https://github.com/jhipster/generator-jhipster/issues/9831 was related to that.

Further, on the keycloak side, our custom `jhipster` client scope is not enabled for access token claims (enabled for user info). As I pointed out previously in https://github.com/jhipster/generator-jhipster/pull/9635, realm level and client level roles are now populated by default in the access token claims. Ideally, we should move to that to be more standard complaint, but, that can be tracked in a separate PR.

If no one is working on it, then, I will try to do PR.
